emscripten.py

```diff
@@ -1407,77 +1447,55 @@ def float_coerce(s):
-       first_in_asm,
-       '''
-  var HEAP8 = new global%s(buffer);
-  var HEAP16 = new global%s(buffer);
-  var HEAP32 = new global%s(buffer);
-  var HEAPU8 = new global%s(buffer);
-  var HEAPU16 = new global%s(buffer);
-  var HEAPU32 = new global%s(buffer);
-  var HEAPF32 = new global%s(buffer);
-  var HEAPF64 = new global%s(buffer);
-''' % (access_quote('Int8Array'),
-     access_quote('Int16Array'),
-     access_quote('Int32Array'),
-     access_quote('Uint8Array'),
-     access_quote('Uint16Array'),
-     access_quote('Uint32Array'),
-     access_quote('Float32Array'),
-     access_quote('Float64Array'))
-     if not settings['ALLOW_MEMORY_GROWTH'] else '''
-  var Int8View = global%s;
-  var Int16View = global%s;
-  var Int32View = global%s;
-  var Uint8View = global%s;
-  var Uint16View = global%s;
-  var Uint32View = global%s;
-  var Float32View = global%s;
-  var Float64View = global%s;
-  var HEAP8 = new Int8View(buffer);
-  var HEAP16 = new Int16View(buffer);
-  var HEAP32 = new Int32View(buffer);
-  var HEAPU8 = new Uint8View(buffer);
-  var HEAPU16 = new Uint16View(buffer);
-  var HEAPU32 = new Uint32View(buffer);
-  var HEAPF32 = new Float32View(buffer);
-  var HEAPF64 = new Float64View(buffer);
-  var byteLength = global.byteLength;
-''' % (access_quote('Int8Array'),
-     access_quote('Int16Array'),
-     access_quote('Int32Array'),
-     access_quote('Uint8Array'),
-     access_quote('Uint16Array'),
-     access_quote('Uint32Array'),
-     access_quote('Float32Array'),
-     access_quote('Float64Array'))) + '\n' + asm_global_vars + ('''
+
+def create_asm_start_pre(asm_setup, the_global, sending, metadata, settings):
+  access_quote = access_quoter(settings)
+
+  shared_array_buffer = ''
+  if settings['USE_PTHREADS']:
+    shared_array_buffer = "Module.asmGlobalArg['Atomics'] = Atomics;"
+
+  module_get = 'Module{access} = {val};'
+  module_global = module_get.format(access=access_quote('asmGlobalArg'), val=the_global)
+  module_library = module_get.format(access=access_quote('asmLibraryArg'), val=sending)
+
+  asm_function_top = ('// EMSCRIPTEN_START_ASM\n'
+                      'var asm = (function(global, env, buffer) {')
+
+  use_asm = "'almost asm';"
+  if not metadata.get('hasInlineJS') and settings['ASM_JS'] == 1:
+    use_asm = "'use asm';"
+
+  lines = [
+    asm_setup,
+    module_global,
+    shared_array_buffer,
+    module_library,
+    asm_function_top,
+    use_asm,
+    create_first_in_asm(settings),
+    create_memory_views(settings),
+  ]
+  return '\n'.join(lines)
+
+
+def create_asm_temp_vars(settings):
+  access_quote = access_quoter(settings)
+  return '''
   var __THREW__ = 0;
   var threwValue = 0;
   var setjmpId = 0;
   var undef = 0;
   var nan = global%s, inf = global%s;
   var tempInt = 0, tempBigInt = 0, tempBigIntP = 0, tempBigIntS = 0, tempBigIntR = 0.0, tempBigIntI = 0, tempBigIntD = 0, tempValue = 0, tempDouble = 0.0;
```

Comment:
They were a hack for handling i64s 2 backends earlier. Almost certainly they can be removed now...